File: container2/design_patterns/creational/factory-method/factory-method.py
from __future__ import annotations

class Application():
    """Creator - Interface"""
    documents = []

    def create_document(self) -> Document:
        return Document()

    def new_document(self):
        doc = self.create_document()
        self.documents.append(doc)
        print("Saved new document")

    def open_document(self):
        pass

class Document():
    """Product - Interface"""

    def open(self):
        pass

    def close(self):
        pass

    def save(self):
        pass

    def revert(self):
        pass

class MyApplication(Application):
    """Concrete Creator"""

    def create_document(self) -> MyDocument:
        return MyDocument()

class MyDocument(Document):
    """Concrete Product"""

    def open(self):
        return "My application-specific document"

def client_code():
    """ Demo code """

    myCreator = MyApplication()
    myCreator.new_document()
    print(myCreator.documents)

if __name__ == "__main__":
    client_code()

File: container2/design_patterns/creational/factory-method/factory-method.js
/**
 * Creator - Interface
 */
class Application {
    _documents = []

    /**
     * Factory Method
     */
    createDocument = function() {
        return new Document()
    }
    
    newDocument = function() {
        this._documents.push(this.createDocument())
        console.log("Saved new document")
    }

    openDocument = function() {}
}

/**
 * Product - Interface
 */
class Document {
    open = function() {}
    close = function() {}
    save = function() {}
    revert = function() {}
}

/**
 * ConcreteCreator
 */
class MyApplication extends Application {
    /**
     * Factory Method for application's specific product
     */
    createDocument = function() {
        return new MyDocument()
    }
}

/**
 * ConcreteProduct
 */
class MyDocument extends Document {
    open = function() { console.log("Opened MyApplication Specific Document") }
}

/**
 * EXAMPLE CODE
 */
function demo() {
    const myCreator = new MyApplication()
    myCreator.newDocument()
    myCreator._documents[0].open()
}

demo()
